// botframework-config/lib/models/luisService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */
const schema_1 = require("../schema");
const connectedService_1 = require("./connectedService");
class LuisService extends connectedService_1.ConnectedService {
    constructor(source = {}, serviceType) {
        super(source, serviceType || schema_1.ServiceTypes.Luis);
    }
    // get endpoint for the luis service
    getEndpoint() {
        return `https://${this.region}.api.cognitive.microsoft.com`;
    }
    // encrypt keys in service
    encrypt(secret, encryptString) {
        if (this.authoringKey && this.authoringKey.length > 0) {
            this.authoringKey = encryptString(this.authoringKey, secret);
        }
        if (this.subscriptionKey && this.subscriptionKey.length > 0) {
            this.subscriptionKey = encryptString(this.subscriptionKey, secret);
        }
    }
    // decrypt keys in service
    decrypt(secret, decryptString) {
        if (this.authoringKey && this.authoringKey.length > 0) {
            this.authoringKey = decryptString(this.authoringKey, secret);
        }
        if (this.subscriptionKey && this.subscriptionKey.length > 0) {
            this.subscriptionKey = decryptString(this.subscriptionKey, secret);
        }
    }
}
exports.LuisService = LuisService;
